src/coarse/coarse.c

```diff
@@ -1053,90 +1071,103 @@ umf_result_t coarse_alloc(coarse_t *coarse, size_t size, size_t alignment,
-        assert(curr->used == false);
-
-        // In case of non-zero alignment create an aligned block what would be further used.
-        if (alignment > 0) {
-            umf_result = create_aligned_block(coarse, size, alignment, &curr);
-            if (umf_result != UMF_RESULT_SUCCESS) {
-                (void)free_blocks_re_add(coarse, curr);
-                goto err_unlock;
-            }
+        if (!coarse->cb.alloc) {
+            LOG_ERR("out of memory");
+            goto err_unlock;
         }
 
-        if (action == ACTION_SPLIT) {
-            // Split the current block and put the new block after the one that we use.
-            umf_result = split_current_block(coarse, curr, size);
-            if (umf_result != UMF_RESULT_SUCCESS) {
-                (void)free_blocks_re_add(coarse, curr);
-                goto err_unlock;
-            }
+        size_t size_aligned = ALIGN_UP_SAFE(size, alignment);
+        if (size_aligned == 0) {
+            // cannot align up (arithmetic overflow)
+            size_aligned = size;
+        }
 
-            curr->size = size;
+        umf_result = coarse->cb.alloc(coarse->provider, size_aligned, alignment,
+                                      resultPtr);
+        if (umf_result != UMF_RESULT_SUCCESS) {
+            LOG_ERR("coarse_alloc_cb() failed: out of memory");
+            goto err_unlock;
+        }
 
-            LOG_DEBUG("coarse_ALLOC (split_block) %zu used %zu alloc %zu", size,
-                      coarse->used_size, coarse->alloc_size);
+        ASSERT_IS_ALIGNED(((uintptr_t)(*resultPtr)), alignment);
 
-        } else { // action == ACTION_USE
-            LOG_DEBUG("coarse_ALLOC (same_block) %zu used %zu alloc %zu", size,
-                      coarse->used_size, coarse->alloc_size);
+        block_t *new_free_block = NULL;
+        umf_result = coarse_add_free_block(coarse, *resultPtr, size_aligned,
+                                           &new_free_block);
+        if (umf_result != UMF_RESULT_SUCCESS) {
+            if (coarse->cb.free) {
+                coarse->cb.free(coarse->provider, *resultPtr, size_aligned);
+            }
+            goto err_unlock;
         }
 
-        curr->used = true;
-        *resultPtr = curr->data;
-        coarse->used_size += size;
+        LOG_DEBUG("coarse_ALLOC (memory_provider) %zu used %zu alloc %zu",
+                  size_aligned, coarse->used_size, coarse->alloc_size);
 
-        assert(debug_check(coarse));
-        utils_mutex_unlock(&coarse->lock);
-
-        return UMF_RESULT_SUCCESS;
+        curr = new_free_block;
     }
 
-    // no suitable block found - try to get more memory from the upstream provider
-    umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
-
-    *resultPtr = NULL;
+    // If the block that we want to reuse has a greater size, split it.
+    // Try to merge the split part with the successor if it is not used.
+    enum { ACTION_NONE = 0, ACTION_USE, ACTION_SPLIT } action = ACTION_NONE;
```

Comment:
`ACTION_NONE`
AFAIK, starting enum entry is `0`, so there is no need to set it up.

Reply:
Right, but IMHO it is more readable as it is now